go/types/sequence_cursor.go

```diff
@@ -4,24 +4,41 @@
 
 package types
 
-import "github.com/attic-labs/noms/go/d"
+import (
+	"math"
+	"os"
+	"runtime"
+
+	"github.com/attic-labs/noms/go/d"
+)
 
 // sequenceCursor explores a tree of sequence items.
 type sequenceCursor struct {
-	parent *sequenceCursor
-	seq    sequence
-	idx    int
+	parent        *sequenceCursor
+	seq           sequence
+	idx           int
+	readAheadMap  map[int]chan sequence
+	readAheadNext int
 }
 
+// newSequenceCursor creates a cursor on seq positioned at idx.
+// If idx < 0, count backward from the end of seq.
 func newSequenceCursor(parent *sequenceCursor, seq sequence, idx int) *sequenceCursor {
 	d.PanicIfTrue(seq == nil)
 	if idx < 0 {
 		idx += seq.seqLen()
 		d.PanicIfFalse(idx >= 0)
 	}
 
-	cur := &sequenceCursor{parent, seq, idx}
-	return cur
+	var readAheadMap map[int]chan sequence
+	// enable read-ahead if the cursor represents to a meta-sequence containing leaves
+	ms, ok := seq.(metaSequence)
+	if os.Getenv("NOMS_NO_READAHEAD") == "" && ok {
+		if ms.tuples[0].ref.height == 1 {
+			readAheadMap = make(map[int]chan sequence)
```

Comment:
Did you think about calling readAhead here immediately? Or is it automatically happening somehow outside the diff of this PR? It certainly seems like after clone() it's not initiated immediately.

It seems reasonable to me that we don't until the first call I suppose, but you could comment about that.